File: src/tableau_cli/errors/cli_error.py
from __future__ import annotations

from typing import Any, Optional


class CliError(Exception):
    def __init__(
        self,
        *,
        error_type: str,
        message: str,
        details: Optional[str] = None,
        hint: Optional[str] = None,
    ):
        super().__init__(message)
        self.error_type = error_type
        self.details = details
        self.hint = hint

    def to_output(self) -> dict[str, Any]:
        output: dict[str, Any] = {
            "isError": True,
            "errorType": self.error_type,
            "message": str(self),
        }
        if self.details:
            output["details"] = self.details
        if self.hint:
            output["hint"] = self.hint
        return output


class FeatureDisabledError(CliError):
    def __init__(self, message: str, hint: Optional[str] = None):
        super().__init__(error_type="feature-disabled", message=message, hint=hint)


class AuthenticationError(CliError):
    def __init__(self, message: str):
        super().__init__(
            error_type="authentication-error",
            message=message,
            hint=(
                "Check your PAT credentials with `tableau-cli config show`. "
                "Ensure the token has not expired."
            ),
        )

File: src/errors/cliError.ts
/**
 * Structured error for CLI output.
 * All errors are output as JSON to stdout so agents can parse them.
 */
export interface CliErrorOutput {
  isError: true;
  errorType: string;
  message: string;
  details?: string;
  hint?: string;
}

export class CliError extends Error {
  readonly errorType: string;
  readonly details?: string;
  readonly hint?: string;

  constructor(opts: { errorType: string; message: string; details?: string; hint?: string }) {
    super(opts.message);
    this.errorType = opts.errorType;
    this.details = opts.details;
    this.hint = opts.hint;
  }

  toOutput(): CliErrorOutput {
    const output: CliErrorOutput = {
      isError: true,
      errorType: this.errorType,
      message: this.message,
    };
    if (this.details) output.details = this.details;
    if (this.hint) output.hint = this.hint;
    return output;
  }
}

export class FeatureDisabledError extends CliError {
  constructor(message: string, hint?: string) {
    super({ errorType: 'feature-disabled', message, hint });
  }
}

export class AuthenticationError extends CliError {
  constructor(message: string) {
    super({
      errorType: 'authentication-error',
      message,
      hint: 'Check your PAT credentials with `tableau-cli config show`. Ensure the token has not expired.',
    });
  }
}
